extension-workspace-references.md

```diff
@@ -0,0 +1,115 @@
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate server-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace references exporting support.
+   */
+  xworkspaceReferencesProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace references request is sent from the client to the server to export project-wide references to dependencies. That is, the response strictly returns references in the project to symbols defined in dependencies.
+
+_Request_
+* method: 'workspace/xreferences'
+* params: `WorkspaceReferencesParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace References Request.
+ */
+interface WorkspaceReferencesParams {
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace where the `symbol` has been referenced.
+	 */
+	reference: Location;
+
+	/**
+	 * Metadata information describing the symbol being referenced.
+	 */
+	symbol: ReferenceSymbolInformation;
+}
+```
+* error: code and message set in case an exception happens during the workspace references request.
+
+Where `ReferenceSymbolInformation` is defined as follows:
+
+```typescript
+/**
+ * Represents information about a programming construct like a variable, class,
+ * interface etc that has a reference to it. Effectively, it contains data similar
+ * to SymbolInformation except all fields are optional and a metadata field is
+ * present for language-specific data.
+ *
+ * ReferenceSymbolInformation does NOT always uniquely identify a symbol. The
+ * caller should effectively consider the returned information to be
+ * information about the symbol which _generally_ (but NOT always) identifies a
+ * single symbol.
+ */
+interface ReferenceSymbolInformation {
+    /**
+     * The name of this symbol (same as `SymbolInformation.name`).
+     */
+    name?: string;
+
+    /**
+     * The kind of this symbol (same as `SymbolInformation.kind`).
+     */
+    kind?: number;
+
+    /**
+     * The URI of this symbol (same as `SymbolInformation.location.uri`).
+     */
+    uri?: string;
+
+    /**
+     * The name of the symbol containing this symbol (same as `SymbolInformation.containerName`).
+     */
+    containerName?: string;
+
+    /**
+     * The repository URI that the symbol is defined in, e.g. `github.com/golang/go`
+     * or `bitbucket.org/jespern/django-piston`.
+     */
+    repo?: string;
```

Comment:
I think that @felixfbecker makes some compelling arguments here.. I will offer my own points:

1. If both Go and PHP have access to repo URL, commit ID, and VCS type (git, svn, etc), then I'd like to see them in the same field not in `metadata.repo` OR `metadata.repository` OR `metadata.github` etc.
2. If Java, Go, PHP all have other things to add (Java may need custom non-repo unique identifier, Go has build tags, PHP has dist information etc.) then they should be able to add it without requiring some new 'generic' definition to match their fields that are semi-language-specific.

The question becomes "which fields do we standardize" versus not.. and I think that isn't an easy question to answer. We may quickly end up with a quite long specification. I think that `metadata` again is the right way to go here.

@felixfbecker I'm going to re-add metadata and remove the `repo` and `packageName` fields. I think we will need `metadata` regardless to handle this. Whether or not we include `repo` or `packageName` I am not sure. If you were to standardize fields like `repo` and `packageName`, would it just be those two? Would you include `commit` or `version` (is that a commit)? Does `repo` have an accompanying VCS type (`git`, `svn`, etc)? I would like to hear your opinion here because I am honestly not sure if that would be best.